mindsdb/integrations/libs/ml_exec_base.py

```diff
@@ -0,0 +1,380 @@
+from collections import OrderedDict
+
+import datetime as dt
+from dateutil.parser import parse as parse_datetime
+import json
+import traceback
+import importlib
+
+import numpy as np
+import pandas as pd
+
+from mindsdb_sql import parse_sql
+from mindsdb_sql.parser.ast.base import ASTNode
+from mindsdb_sql.parser.ast import BinaryOperation, Identifier, Constant, Select, Show, Star, NativeQuery
+from mindsdb_sql.parser.dialects.mindsdb import (
+    RetrainPredictor,
+    CreatePredictor,
+    DropPredictor
+)
+
+
+from mindsdb.integrations.utilities.utils import make_sql_session, get_where_data
+
+from mindsdb.utilities.config import Config
+import mindsdb.interfaces.storage.db as db
+from mindsdb.integrations.libs.response import (
+    HandlerResponse as Response,
+    RESPONSE_TYPE
+)
+from mindsdb import __version__ as mindsdb_version
+from mindsdb.utilities.hooks import after_predict as after_predict_hook
+from mindsdb.utilities.with_kwargs_wrapper import WithKWArgsWrapper
+from mindsdb.interfaces.model.model_controller import ModelController
+from mindsdb.interfaces.model.functions import (
+    get_model_record,
+    get_model_records
+)
+from mindsdb.api.mysql.mysql_proxy.classes.sql_query import SQLQuery
+
+from mindsdb.integrations.libs.const import PREDICTOR_STATUS
+from mindsdb.integrations.utilities.processes import HandlerProcess
+from mindsdb.utilities.functions import mark_process
+from mindsdb.integrations.utilities.utils import format_exception_error
+
+from mindsdb.interfaces.storage.fs import ModelStorage, HandlerStorage
+
+import torch.multiprocessing as mp
+ctx = mp.get_context('spawn')
+
+@mark_process(name='learn')
+def learn_process(class_path, company_id, integration_id, predictor_id, training_data_df, target, problem_definition):
+
+    predictor_record = db.Predictor.query.with_for_update().get(predictor_id)
+
+    predictor_record.training_start_at = dt.datetime.now()
+    db.session.commit()
+
+    try:
+        module_name, class_name = class_path
+        module = importlib.import_module(module_name)
+        klass = getattr(module, class_name)
+
+        handlerStorage = HandlerStorage(company_id, integration_id)
+        modelStorage = ModelStorage(company_id, predictor_id)
+
+        ml_handler = klass(
+            engine_storage=handlerStorage,
+            model_storage=modelStorage,
+        )
+        ml_handler.create(target, df=training_data_df, args=problem_definition)
+
+    except Exception as e:
+        print(traceback.format_exc())
+        error_message = format_exception_error(e)
+
+        predictor_record.data = {"error": error_message}
+        predictor_record.status = PREDICTOR_STATUS.ERROR
+        db.session.commit()
+
+    predictor_record.training_stop_at = dt.datetime.now()
+    predictor_record.status = PREDICTOR_STATUS.COMPLETE
+    db.session.commit()
+
+
+@mark_process(name='predict')
+def predict_process(class_path, company_id, integration_id, predictor_id, df, res_queue=None):
+
```

Comment:
add docs https://google.github.io/styleguide/pyguide.html#s3.8-comments-and-docstrings